docs/internal/git_best_practice.rst

```diff
@@ -0,0 +1,211 @@
+*merged* in any order creating a rather non-linear history which can be hard to
+follow. We adopt a more constrained approach, commonly referred as "Feature
+Branch Workflow". In short this means that there is one main branch ``master``
+or ``develop``, no major work is done in the main branch, instead new features
+are developed in separate feature branches, those then get merged into the main
+branch when ready.
+
+
+Commit Message Format
+-----------------------------
+
+Good commit messages make collaboration easier. Commit message should consist of
+short summary line, 50 characters of less, followed by a blank line, followed by
+expanded description. It is fine to omit expanded description and the blank line
+before it if the change is small. Start your summary line with a capital letter,
+don't put full stop [.] at the end of the summary line.
+
+Many tools, GitHub including, expect this format and sticking to it makes
+inspecting history more accessible.
+
+.. note ::
+
+    Could be helpful to prescribe how summary line is to be phrased. Which do we
+    prefer?
+
+    - Fixed spelling mistake in the help message
+    - Fix spelling mistake in the help message
+
+For a more expanded exploration of the topic see `<https://chris.beams.io/posts/git-commit/>`_
+
+
+Clean Linear History
+---------------------
+
+Pull with Rebase
+^^^^^^^^^^^^^^^^^
+
+When grabbing changes from remote use ``--rebase`` switch, this keeps history
+linear by replaying your changes on top of remote changes.
+
+For example ``git pull --rebase origin master``. Alternatively make pull with
+rebase a global default ``git config --global pull.rebase true``. Your editor
+probably has this option, certainly PyCharm supports pulling with rebase.
+
+
+Cleanup Before Sharing
+^^^^^^^^^^^^^^^^^^^^^^
+
+Just like code can be edited, so can be git history. You don't need to type out
+perfect code on the first try, you don't need to delay committing your code
+until it's "just right", "work in progress" commits are perfectly fine so long
+as you clean it up before sharing your changes.
+
+The easiest and most common method is to *amend* the very last commit ``git
+commit --amend``. Maybe you want to correct a typo in the commit message, or you
+forgot to add a new file, you can easily amend last commit. Your editor most
+likely supports that out of the box.
+
+A more comprehensive method is *Interactive Rebase*, it allows to re-order
+commits, squash independent commits into one, update commit messages for older
+commits and more. While you can use it with just plain command line ``git
+rebase -i HEAD~3``, it's probably best to use some kind of GUI client for
+complex tasks like that. Maybe your editor/IDE already supports that, if not try
+some of the tools listed below
+
+- `<https://www.gitkraken.com>`_
+- `<https://www.sourcetreeapp.com>`_
+- Docs for plain git `<https://help.github.com/articles/changing-a-commit-message/>`_
+
+
+Initial Setup
```

Comment:
Maybe move this "setup" section to the top?

Reply:
yeah, it does feel out of the place, will do